spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,689 @@
+
+According to the Tendermint algorithm, a correct process can only broadcast one precommit message in round $r$, height $h$.
+Since, as stated in the [Description](#description) section, `ResponseExtendVote` is only called when Tendermint
+is about to broadcast a non-`nil` precommit message, a correct process can only produce one vote extension in round $r$, height $h$.
+Let $e^r_p$ the vote extension that the Application of a correct process $p$ returns via `ResponseExtendVote` in round $r$, height $h$.
+Let $w^r_p$ the proposed block that $p$'s Tendermint passes to the Application via `RequestExtendVote` in round $r$, height $h$.
+
+* Property 6 [`ExtendVote`, `VerifyVoteExtension`, coherence]: For any two correct processes $p$ and $q$, if $q$ receives $e^r_p$
+  from $p$ in height $h$, $q$'s Application returns Accept in `ResponseVerifyVoteExtension`.
+
+Property 6 constrains the creation and handling of vote extensions in a similar way as Property 3
+contrains the creation and handling of proposed blocks.
+Property 6 ensures that extensions created by correct processes _always_ pass the `VerifyVoteExtension`
+checks performed by correct processes receiving those extensions.
+However, if there is a (deterministic) bug in `ExtendVote` or `VerifyVoteExtension` (or in both),
+we will face the same liveness issues as described for Property 3, as Precommit messages with invalid vote
+extensions will be discarded.
+
+* Property 7 [`VerifyVoteExtension`, determinism-1]: For any correct process $p$,
+  and any arbitrary vote extension $e$, and any arbitrary block $w$,
+  if $p$'s (resp. $q$'s) Tendermint calls `RequestVerifyVoteExtension` on $e$ and $w$ at height $h$,
+  then $p$'s Application's acceptance or rejection exclusively depends on $e$, $w$ and $s_{p,h-1}$.
+
+* Property 8 [`VerifyVoteExtension`, determinism-2]: For any two correct processes $p$ and $q$,
+  and any arbitrary vote extension $e$, and any arbitrary block $w$,
+  if $p$'s (resp. $q$'s) Tendermint calls `RequestVerifyVoteExtension` on $e$ and $w$ at height $h$,
+  then $p$'s Application accepts $e$ if and only if $q$'s Application accepts $e$.
+  Note that this property follows from Property 7 and the Agreement property of consensus.
+
+Properties 7 and 8 ensure that the validation of vote extensions will be deterministic at all
+correct processes.
+Properties 7 and 8 protect against arbitrary vote extension data from Byzantine processes
+similarly to Properties 4 and 5 and proposed blocks.
+Properties 7 and 8 can be violated by a bug inducing non-determinism in `ExtendVote` or
+`VerifyVoteExtension`. In this case liveness can be compromised.
+Extra care should be put in the implementation of `ExtendVote` and `VerifyVoteExtension` and,
+as a general rule, `VerifyVoteExtension` _should_ always accept the vote extension.
+
+* Property 9 [_all_, read-only]: $p$'s calls to `RequestPrepareProposal`, `RequestProcessProposal`,
```

Comment:
Should we say "free of side effects" rather than "read-only"?

Reply:
Yep. Look reasonable